src/main/java/org/logstash/beats/Server.java

```diff
@@ -83,13 +87,30 @@ public void stop() {
     }
 
     private void shutdown() {
+        // as much as possible we try to gracefully shut down
+        // no longer accept incoming socket connections
+        // with event loop group (workGroup) declares quite period with `shutdownGracefully` which queued tasks will not receive work for the best cases
+        // executor group threads will be asked and waited to gracefully shutdown if they have pending tasks. This helps each individual handler process the event/exception, especially when multichannel use case.
+        //  there is no guarantee that executor threads will terminate during the shutdown because of many factors such as ack processing
+        //  so any pending tasks which send batches to BeatsHandler will be ignored
         try {
+            // boss group shuts down socket connections
+            // shutting down bossGroup separately gives us faster channel closure
+            if (bossGroup != null) {
+                bossGroup.shutdownGracefully().sync();
+            }
+
             if (workGroup != null) {
-                workGroup.shutdownGracefully().sync();
+                workGroup.shutdownGracefully();
             }
             if (beatsInitializer != null) {
                 beatsInitializer.shutdownEventExecutor();
             }
+            // we wait for workGroup shutdown this time
+            if (workGroup != null && !workGroup.isShutdown()) {
+                // calling again shutdownGracefully doesn't hurt that it returns terminationFuture if shutdown is under process
+                workGroup.shutdownGracefully().sync();
+            }
```

Comment:
Not clear why after  the previous call `workGroup.shutdownGracefully` we need a second one.

`shutdownGracefully` at first thing sets its status in `ST_SHUTTING_DOWN`.  Maybe it's due to the default quiet period (2 seconds by default), so that is could happen that after the first call to `workGroup.shutdownGracefully` another task fall in the task queue before the quiet period elapses, and the shutdown status is reset.  If this is the case, we could set 0 as quiet period or `sync` on the first call.

Reply:
Yup, thanks for catching up, forgot to push my last local update.
Worker shutdown doesn't need to necessarily be in this order, simply boss -> group -> event executors termination is a technically correct way. I have pushed the change.